### ISSUE TITLE
Infinite Loop in React 18+

### ISSUE THREAD

**Issue by sum117** [keycloak/keycloak#19452]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

adapter/javascript

### Describe the bug

If you are using the JavaScript Adapter to create a KeycloakProvider instance, or even using the react component for it (I tried with both) you'll create an infinite loop UNLESS you remove strict mode OR place the provider outside the StrictMode.

This shouldn't be the right way to do it. It's caused by the keycloak.init method.

### Version

latest

### Expected behavior

You should be able to use it without it causing re-renders WITH the adapter provider inside the strictMode

### Actual behavior

Calling the init function with a built provider anywhere in the application or just by using the adapter in a login page will trigger an infinite loop of re-renders WHILE using react Strict Mode in 18+

### How to Reproduce?

Make a provider with a context or simply a hook that uses the .init() method and place it inside strict mode. While trying to trigger "check-sso", Keycloak will infinite loop until you remove strict mode. You can kind of fix this with top level await on the .init() method but this IS NOT SUPPORTED BY MOST BROWSERS.

### Anything else?

https://github.com/react-keycloak/react-keycloak/issues/182

It's been almost a year...

**Comment by jonkoops** [comment]
This is an issue in React Keycloak, not Keycloak JS. Likely it is calling methods of Keycloak JS that should only be called once. Since this is not an issue on our side I am closing this ticket.

**Comment by sum117** [comment]
Sorry for restarting the thread with another message, but what I meant is that I wasn't using React Keycloak (although this is an issue that they have in their package too). I was literally making my own custom provider, and I'm confident it's not a problem with it since it works when I get it out of react's strict mode. I don't mind sharing:

```js

import React, { type ReactNode } from "react";
import Keycloak from "keycloak-js";
import LoadingPage from "components/pages/LoadingPage";

interface KeycloakContextProps {
  keycloak: Keycloak;
}

const KeycloakContext = React.createContext<KeycloakContextProps | null>(null);

export const useKeycloak = (): KeycloakContextProps => {
  const context = React.useContext(KeycloakContext);

  if (context === null) {
    throw new Error("useKeycloak must be used within a KeycloakProvider");
  }

  return context;
};

const instance = new Keycloak({
  clientId: "my-client-id",
  realm: "master",
  url: "http://localhost:8080/",
});

const useKeycloakState = (): boolean => {
  const [isLoading, setIsLoading] = React.useState(true);

  React.useEffect(() => {
    instance
      .init({
        onLoad: "check-sso",
        pkceMethod: "S256",
        scope: "openid profile email",
        redirectUri: "http://localhost:4173/login",
        checkLoginIframe: false,
      })
      .then(() => {
        setIsLoading(false);
      })
      .catch(console.error);
  }, []);

  return isLoading;
};

export default function KeycloakProvider({
  children,
}: {
  children: ReactNode;
}): JSX.Element {
  const isLoading = useKeycloakState();
  return (
    <KeycloakContext.Provider value={{ keycloak: instance }}>
      {!isLoading ? children : <LoadingPage />}
    </KeycloakContext.Provider>
  );
}
```

**Comment by sum117** [comment]
@jonkoops

**Comment by jonkoops** [comment]
```js
useEffect(() => {
  instance.init({
    onLoad: "check-sso",
    pkceMethod: "S256",
    scope: "openid profile email",
    redirectUri: "http://localhost:4173/login",
    checkLoginIframe: false,
  });
}, []);
```

This will call `init()` twice in strict mode, it can only be called once. According to the [React docs](https://react.dev/reference/react/StrictMode#fixing-bugs-found-by-re-running-effects-in-development):

> When Strict Mode is on, React will also run **one extra setup+cleanup cycle in development for every Effect**. This may feel surprising, but it helps reveal subtle bugs that are hard to catch manually.

I'd recommend keeping track of whether you already initialized (untested):

```js
const didInit = useRef(false);

useEffect(() => {
  if (didInit.current) {
    return;
  }

  didInit.current = true;
  instance.init();
}, []);
```

**Comment by sum117** [comment]
Will test. Thank you for the advice.
________________________________
De: Jon Koops ***@***.***>
Enviado: terça-feira, 11 de abril de 2023 11:33
Para: keycloak/keycloak ***@***.***>
Cc: sum117 ***@***.***>; Author ***@***.***>
Assunto: Re: [keycloak/keycloak] Infinite Loop in React 18+ (Issue #19452)


useEffect(() => {
  instance.init({
    onLoad: "check-sso",
    pkceMethod: "S256",
    scope: "openid profile email",
    redirectUri: "http://localhost:4173/login",
    checkLoginIframe: false,
  });
}, []);

This will call init() twice in strict mode, it can only be called once. According to the React docs<https://react.dev/reference/react/StrictMode#fixing-bugs-found-by-re-running-effects-in-development>:

When Strict Mode is on, React will also run one extra setup+cleanup cycle in development for every Effect. This may feel surprising, but it helps reveal subtle bugs that are hard to catch manually.

I'd recommend keeping track of whether you already initialized (untested):

const didInit = useRef(false);

useEffect(() => {
  if (didInit.current) {
    return;
  }

  didInit.current = true;
  instance.init();
}, []);

**Comment by davoodhobabi** [comment]
I still have this issue even when I remove strict mode at all

I created a singleton of instances of the Keycloak and I'm using it in my App.js file

**Comment by shighetari** [comment]
> I still have this issue even when I remove strict mode (React v18.2.0)
> 
> 
> 
> I created a singleton of instances of the Keycloak and I'm using it in my App.js file 

I'm also facing a similar delima. It seems that this is an issue with react however, and not the keycloak team from what I am understanding. I'm going to attempt to use the useRef as a way to track the results of authentication as apposed to calling another instance of keycloak. The double rendering due to strict mode in development just made me realize I spent most of my day today growing more frustrated when all I had to do was spend half an hour reading thru googled results..
